daemon/smartcard.c

```diff
@@ -360,6 +744,98 @@ smartcard_change_pin(smartcard_t *smartcard, control_t *control, const char *pas
 	return (ret > 0) ? 0 : -1;
 }
 
+/**
+ * apparently we cannot queue several events with the same fd.
+ * therefore, we use a blocking method to query the scd to initialize a token.
+ */
+int
+smartcard_scd_token_block_add(smartcard_t *smartcard, container_t *container)
+{
+	TRACE("CML: smartcard_scd_token_block_new");
+	ASSERT(smartcard);
+	ASSERT(container);
+
+	DaemonToToken out = DAEMON_TO_TOKEN__INIT;
+	out.code = DAEMON_TO_TOKEN__CODE__TOKEN_ADD;
+
+	out.token_uuid = mem_strdup(uuid_string(container_get_uuid(container)));
+
+	out.has_token_type = true;
+	out.token_type = smartcard_tokentype_to_proto(container_get_token_type(container));
+
+	if (out.token_type == TOKEN_TYPE__USB) {
+		out.usbtoken_serial = container_get_usbtoken_serial(container);
+		if (NULL == out.usbtoken_serial) {
+			ERROR("Could not retrive serial os usbtoken reader. Abort token init...");
+			mem_free(out.token_uuid);
+			return -1;
+		}
+	}
+
+	TokenToDaemon *msg = smartcard_send_recv_block(&out);
+	if (!msg) {
+		ERROR("Failed to receive message although EVENT_IO_READ was set. Aborting smartcard_scd_token_block_new.");
+		return -1;
+	}
+
+	switch (msg->code) {
+	case TOKEN_TO_DAEMON__CODE__TOKEN_ADD_SUCCESSFUL: {
+		TRACE("CMLD: smartcard_scd_token_block_new: token in scd created successfully");
+		container_set_token_uuid(container, out.token_uuid);
+		container_set_token_is_init(container, true);
+	} break;
+	case TOKEN_TO_DAEMON__CODE__TOKEN_ADD_FAILED: {
+		container_set_token_is_init(container, false);
+		ERROR("Creating scd token structure failed");
+	} break;
+	default:
+		container_set_token_is_init(container, false);
+		ERROR("TokenToDaemon command %d not expected as answer to change_pin", msg->code);
+	}
+
+	mem_free(out.token_uuid);
+	protobuf_free_message((ProtobufCMessage *)msg);
+	return 0;
+}
+
+int
+smartcard_scd_token_block_remove(smartcard_t *smartcard, container_t *container)
+{
+	TRACE("CML: smartcard_scd_token_block_remove");
+	ASSERT(smartcard);
+	ASSERT(container);
+
+	DaemonToToken out = DAEMON_TO_TOKEN__INIT;
+	out.code = DAEMON_TO_TOKEN__CODE__TOKEN_REMOVE;
+
+	out.token_uuid = mem_strdup(uuid_string(container_get_uuid(container)));
+
+	out.has_token_type = true;
+	out.token_type = smartcard_tokentype_to_proto(container_get_token_type(container));
+
+	TokenToDaemon *msg = smartcard_send_recv_block(&out);
+	if (!msg) {
+		ERROR("Failed to receive message although EVENT_IO_READ was set. Aborting smartcard_scd_token_block_new.");
+		return -1;
```

Comment:
goto error + free token_uuid

Reply:
freed uuid explicitly as this is the only error case in this function